gremlin/query.go

```diff
@@ -32,90 +32,170 @@ import (
 // QueryString used to construct string representation of query
 type QueryString string
 
+// NewValueStringFromArgument via inferance creates a correct ValueString
+func NewQueryStringFromArgument(v interface{}) QueryString {
+	switch t := v.(type) {
+	case QueryString:
+		return t
+	case string:
+		return QueryString(t)
+	case fmt.Stringer:
+		return QueryString(t.String())
+	default:
+		panic(fmt.Sprintf("argument %v: type %T not supported", t, v))
+	}
+}
+
 // String converts value to string
 func (v QueryString) String() string {
-	return string(v)
+	s := string(v)
+	fmt.Printf("Gremlin: %s\n", s)
+	return s
 }
 
 // G represents the base graph token
 const G = QueryString("G")
 
 // Append appends string value to query
-func (q QueryString) Append(s string) QueryString {
-	return QueryString(q.String() + s)
+func (q QueryString) appends(s string) QueryString {
+	return QueryString(string(q) + s)
 }
 
-// Append a Context() operation to query
-func (q QueryString) Context(t time.Time) QueryString {
-	if !t.IsZero() {
-		return q.Append(fmt.Sprintf(".Context(%d)", common.UnixMillis(t)))
+func (q QueryString) newQueryString(name string, list ...interface{}) QueryString {
+	q = q.appends(fmt.Sprintf(".%s(", name))
+	first := true
+	for _, v := range list {
+		if !first {
+			q = q.appends(", ")
+		}
+		first = false
+		q = q.appends(string(NewValueStringFromArgument(v)))
 	}
-	return q
+	return q.appends(")")
 }
 
-// V append a V() operation to query
-func (q QueryString) V() QueryString {
-	return q.Append(".V()")
+// Aggregates append a Aggregates() operation to query
+func (q QueryString) Aggregates() QueryString {
+	return q.newQueryString("Aggregates")
 }
 
-// Has append a Has() operation to query
-func (q QueryString) Has(list ...ValueString) QueryString {
-	q = q.Append(".Has(")
+// At append a At() operation to query
+func (q QueryString) At(list ...interface{}) QueryString {
+	return q.newQueryString("At", list...)
+}
+
+// Both append a Both() operation to query
+func (q QueryString) Both() QueryString {
+	return q.newQueryString("Both")
+}
+
+// BPF append a PBF() operation to query
+func (q QueryString) BPF(list ...interface{}) QueryString {
+	return q.newQueryString("BPF", list...)
+}
+
+// CaptureNode append a CaptureNode() operation to query
+func (q QueryString) CaptureNode() QueryString {
+	return q.newQueryString("CaptureNode")
+}
+
+// Context append a Context() operation to query
+func (q QueryString) Context(list ...interface{}) QueryString {
+	newQ := q.appends(".Context(")
 	first := true
 	for _, v := range list {
 		if !first {
-			q = q.Append(", ")
+			newQ = newQ.appends(", ")
+		}
+		switch t := v.(type) {
+		case time.Time:
+			if t.IsZero() {
+				return q
+			}
+			newQ = newQ.appends(fmt.Sprintf("%d", common.UnixMillis(t)))
+		case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
+			if t == -1 {
```

Comment:
That can not work, see https://stackoverflow.com/questions/40575033/golang-multiple-case-in-type-switch

I suggest we don't support -1 as a parameter to Context and do instead:

```
diff --git a/tests/flow_test.go b/tests/flow_test.go
index 534c33fd..e64616c4 100644
--- a/tests/flow_test.go
+++ b/tests/flow_test.go
@@ -495,7 +495,12 @@ func TestFlowGremlin(t *testing.T) {
 }
 
 func queryFlowMetrics(gh *gclient.GremlinQueryHelper, bridge string, timeContext int64, pings int64) error {
-       ovsGremlin := g.G.Context(timeContext).V().Has("Name", bridge, "Type", "ovsbridge")
+       ovsGremlin := g.G
+       if timeContext == -1 {
+               ovsGremlin = ovsGremlin.Context(timeContext)
+       }
+
+       ovsGremlin = ovsGremlin.V().Has("Name", bridge, "Type", "ovsbridge")
        if _, err := gh.GetNode(ovsGremlin); err != nil {
                return err
        }
```